witness/golang/omniwitness/internal/omniwitness/omniwitness.go

```diff
@@ -0,0 +1,238 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// omniwitness provides a single Main file that basically runs the omniwitness.
+// Some components are left pluggable so this can be deployed on different
+// runtimes.
+package omniwitness
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/google/trillian-examples/formats/log"
+	"github.com/google/trillian-examples/serverless/config"
+	wimpl "github.com/google/trillian-examples/witness/golang/cmd/witness/impl"
+	"github.com/google/trillian-examples/witness/golang/internal/persistence/inmemory"
+	"github.com/google/trillian-examples/witness/golang/internal/witness"
+	"github.com/google/trillian-examples/witness/golang/omniwitness"
+	"github.com/gorilla/mux"
+	"golang.org/x/mod/sumdb/note"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"gopkg.in/yaml.v3"
+
+	"github.com/google/trillian-examples/internal/feeder"
+	"github.com/google/trillian-examples/internal/feeder/pixelbt"
+	"github.com/google/trillian-examples/internal/feeder/rekor"
+	"github.com/google/trillian-examples/internal/feeder/serverless"
+	"github.com/google/trillian-examples/internal/feeder/sumdb"
+)
+
+const (
+	// Interval between attempts to feed checkpoints
+	feedInterval = 5 * time.Minute
+)
+
+// singleLogFeederConfig encapsulates the feeder config for a feeder that can only
+// feed a single log.
+type singleLogFeederConfig struct {
+	// Log defines the source log to feed from.
+	Log config.Log `yaml:"Log"`
+}
+
+// multiLogFeederConfig encapsulates the feeder config for a feeder that can only
+// feed a single log.
+// TODO(mhutchinson): why do we do this to ourselves with similar but different configs?!
+// See if we can standardize on them all being plural.
+type multiLogFeederConfig struct {
+	// Log defines the source log to feed from.
+	Logs []config.Log `yaml:"Logs"`
+}
+
+// Main runs the omniwitness, with the witness listening using the listener, and all
+// outbound HTTP calls using the client provided.
+func Main(ctx context.Context, signer note.Signer, httpListener net.Listener, httpClient *http.Client) error {
+	// This error group will be used to run all top level processes
+	g := errgroup.Group{}
+
+	type logFeeder func(context.Context, config.Log, feeder.Witness, *http.Client, time.Duration) error
+	feeders := make(map[config.Log]logFeeder)
+	// Feeder: SumDB
+	sumdbConfig := config.Log{
+		Origin:    "go.sum database tree",
+		PublicKey: "sum.golang.org+033de0ae+Ac4zctda0e5eza+HJyk9SxEdh+s3Ux18htTTAD8OuAn8",
+	}
```

Comment:
Feels like this should be in the config file too, so it's not a bit different/special?